### ISSUE TITLE
HDF-4.2.15-GCCcore-10.2.0 fails to build on aarch64, patching failed

### ISSUE THREAD

**Issue by rungitta** [easybuilders/easybuild-easyconfigs#12786]
In trying to install `R-4.0.4-foss-2020b` on our ARM-server (graviton, on premise), the dependency `HDF-4.2.15-GCCcore-10.2.0` does not install. It fails with this message:

```bash
== building...
== FAILED: Installation ended unsuccessfully (build directory: /opt/software/easybuild/build/HDF/4.2.15/GCCcore-10.2.0): build failed (first 300 chars): cmd " make -j 64 " exited with exit code 2 and output:
Making all in hdf
make[1]: Entering directory '/opt/software/easybuild/build/HDF/4.2.15/GCCcore-10.2.0/hdf-4.2.15/hdf'
Making all in src
make[2]: Entering directory '/opt/software/easybuild/build/HDF/4.2.15/GCCcore-10.2.0/hdf-4.2.15/hdf/src'
mak (took 25 sec)
```

A bunch of errors looking like these come up in the logfile:

```bash
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I/opt/software/easybuild/software/libtirpc/1.3.1-GCCcore-10.2.0/include -I/opt/software/easybuild/software/zlib/1.2.11-GCCcore-1
In file included from hdf.h:20,
                 from dfgr.c:48:
hdfi.h:1119:1: error: unknown type name No
 1119 | No machine type has been defined.  Your Makefile needs to have someing like
      | ^~
```

Getting some advice from @terjekv I tried to patch the ebconfig with patches referred to here:
https://github.com/NixOS/nixpkgs/commit/c1a2918c849a5bc766c6d55d96bc6cf85c9d27f4#annotation_776560839

Patching seems to go ok, and subsequent --dry-run goes well (shows all deps already installed). But when I then try to install I get this:  

```bash
== patching...
== FAILED: Installation ended unsuccessfully (build directory: /home/palrosen/.local/easybuild/build/HDF/4.2.15/GCCcore-10.2.0): build failed (first 300 chars): Couldn't apply patch file /home/palrosen/hdf-patchwork/hdf-aarch64.patch. Process exited with code 1: patching file hdf/src/hconv.h
Hunk #1 FAILED at 59.
1 out of 1 hunk FAILED -- saving rejects to file hdf/src/hconv.h.rej
patching file hdf/src/hdfi.h
Hunk #1 FAILED at 79.
Hunk #2 succeeded at 1115  (took 7 sec)
== Results of the build can be found in the log file(s) /tmp/eb-7_l176ir/easybuild-HDF-4.2.15-20210430.091440.eLbXO.log
ERROR: Build of /home/palrosen/hdf-patchwork/HDF-4.2.15-GCCcore-10.2.0.eb failed (err: "build failed (first 300 chars): Couldn't apply patch file /home/palrosen/hdf-patchwork/hdf-aarch64.patch. Process exited with code 1: patching file hdf/src/hconv.h\nHunk #1 FAILED at 59.\n1 out of 1 hunk FAILED -- saving rejects to file hdf/src/hconv.h.rej\npatching file hdf/src/hdfi.h\nHunk #1 FAILED at 79.\nHunk #2 succeeded at 1115 ")
```
The logfile shows the error starts here:

```bash
== 2021-04-30 09:16:24,310 easyblock.py:2086 INFO Applying patch hdf-aarch64.patch
== 2021-04-30 09:16:24,316 run.py:225 INFO running cmd: patch -b -p1 -i /home/palrosen/hdf-patchwork/hdf-aarch64.patch
== 2021-04-30 09:16:24,463 run.py:595 INFO parse_log_for_error msg: Command used: patch -b -p1 -i /home/palrosen/hdf-patchwork/hdf-aarch64.patch
== 2021-04-30 09:16:24,464 run.py:597 INFO parse_log_for_error (some may be harmless) regExp (?<![(,-]|\w)(?:error|segmentation fault|failed)(?![(,-]|\.?\w) found:
Hunk #1 FAILED at 59.
1 out of 1 hunk FAILED -- saving rejects to file hdf/src/hconv.h.rej
Hunk #1 FAILED at 79.
1 out of 2 hunks FAILED -- saving rejects to file hdf/src/hdfi.h.rej
Hunk #1 FAILED at 34.
Hunk #2 FAILED at 65.
Hunk #3 FAILED at 420.
Hunk #4 FAILED at 908.
Hunk #5 FAILED at 1022.
Hunk #6 FAILED at 1133.
Hunk #7 FAILED at 1190.
Hunk #8 FAILED at 1326.
Hunk #9 FAILED at 1468.
Hunk #10 FAILED at 1606.
Hunk #11 FAILED at 1677.
Hunk #12 FAILED at 1843.
Hunk #13 FAILED at 2008.
13 out of 13 hunks FAILED -- saving rejects to file mfhdf/fortran/jackets.c.in.rej
Hunk #1 FAILED at 293.
1 out of 1 hunk FAILED -- saving rejects to file mfhdf/libsrc/netcdf.h.in.rej
```

`hconv.h.rej` contains this:

```bash
--- hdf/src/hconv.h.aarch64 2020-04-30 21:16:25.325499508 -0600
+++ hdf/src/hconv.h 2020-04-30 21:16:25.338499580 -0600
@@ -59,7 +59,7 @@
 /* CONSTANT DEFINITIONS                                                      */
 /*****************************************************************************/
 /* Generally Big-Endian machines */
-#if !defined(INTEL86) && !defined(MIPSEL) && !defined(DEC_ALPHA) && !defined(I860) && !defined(SUN386) && !(defined(__ia64) && !(defined(hpux) || defined(__hpux))) && !def
+#if !defined(INTEL86) && !defined(MIPSEL) && !defined(DEC_ALPHA) && !defined(I860) && !defined(SUN386) && !(defined(__ia64) && !(defined(hpux) || defined(__hpux))) && !def
 #       define UI8_IN     DFKnb1b   /* Unsigned Integer, 8 bits */
 #       define UI8_OUT    DFKnb1b
 #       define SI16_IN    DFKnb2b   /* S = Signed */
```
Anyone seen this before and knows what I should do?

**Comment by rungitta** [comment]
Next time I'll make sure to tread the instructions a little more carefully:
```  # Note that the PPC, SPARC and s390 patches are only needed so the aarch64 patch applies cleanly```

So the order of the patching was important and it fixed the issue. HDF builds and I'm closing the issue.

**Comment by boegel** [comment]
Re-opening this, as a reminder to add a proper patch to fix the installation of HDF 4.2.15 on aarch64...
